File: lld/problems/kafka-lld/go/broker.go
package kafkalld

import (
	"errors"
	"sync"
)

var (
	ErrTopicExists       = errors.New("topic already exists")
	ErrTopicNotFound     = errors.New("topic not found")
	ErrPartitionNotFound = errors.New("partition not found")
)

// Broker is the top-level orchestrator: it owns topics and consumer groups.
type Broker struct {
	mu             sync.RWMutex
	topics         map[string]*Topic
	consumerGroups map[string]*ConsumerGroup
}

func NewBroker() *Broker {
	return &Broker{
		topics:         make(map[string]*Topic),
		consumerGroups: make(map[string]*ConsumerGroup),
	}
}

func (b *Broker) CreateTopic(name string, numPartitions int) error {
	b.mu.Lock()
	defer b.mu.Unlock()
	if _, exists := b.topics[name]; exists {
		return ErrTopicExists
	}
	topic, err := NewTopic(name, numPartitions, &KeyHashPartitioner{})
	if err != nil {
		return err
	}
	b.topics[name] = topic
	return nil
}

func (b *Broker) Produce(topicName, key, value string) (partitionID int, offset int64, err error) {
	b.mu.RLock()
	topic, ok := b.topics[topicName]
	b.mu.RUnlock()
	if !ok {
		return 0, 0, ErrTopicNotFound
	}
	partitionID = topic.SelectPartition(key)
	offset = topic.Partitions[partitionID].Append(key, value)
	return partitionID, offset, nil
}

func (b *Broker) getOrCreateConsumerGroup(groupID string) *ConsumerGroup {
	b.mu.Lock()
	defer b.mu.Unlock()
	g, ok := b.consumerGroups[groupID]
	if !ok {
		g = NewConsumerGroup(groupID)
		b.consumerGroups[groupID] = g
	}
	return g
}

// Consume returns up to maxMessages messages after the group's last
// committed offset for (topic, partitionID), then auto-commits past the
// returned batch. Auto-commit-on-read keeps this exercise's API small and
// deterministic; see the README trade-offs section for the alternative
// (explicit CommitOffset, at-least-once redelivery on crash).
func (b *Broker) Consume(groupID, topicName string, partitionID, maxMessages int) ([]Message, error) {
	b.mu.RLock()
	topic, ok := b.topics[topicName]
	b.mu.RUnlock()
	if !ok {
		return nil, ErrTopicNotFound
	}
	if partitionID < 0 || partitionID >= len(topic.Partitions) {
		return nil, ErrPartitionNotFound
	}

	group := b.getOrCreateConsumerGroup(groupID)
	fromOffset := group.committedOffset(topicName, partitionID)
	messages := topic.Partitions[partitionID].Read(fromOffset, maxMessages)
	if len(messages) > 0 {
		group.commit(topicName, partitionID, messages[len(messages)-1].Offset+1)
	}
	return messages, nil
}

func (b *Broker) CommittedOffset(groupID, topicName string, partitionID int) int64 {
	group := b.getOrCreateConsumerGroup(groupID)
	return group.committedOffset(topicName, partitionID)
}

File: lld/problems/kafka-lld/java/src/Broker.java
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/** Top-level orchestrator: it owns topics and consumer groups. */
public class Broker {
    public static class TopicExistsException extends RuntimeException {
        public TopicExistsException() {
            super("topic already exists");
        }
    }

    public static class TopicNotFoundException extends RuntimeException {
        public TopicNotFoundException() {
            super("topic not found");
        }
    }

    public static class PartitionNotFoundException extends RuntimeException {
        public PartitionNotFoundException() {
            super("partition not found");
        }
    }

    /** Result of a produce call: the partition chosen and the offset assigned. */
    public static final class ProduceResult {
        public final int partitionId;
        public final long offset;

        public ProduceResult(int partitionId, long offset) {
            this.partitionId = partitionId;
            this.offset = offset;
        }
    }

    private final ReadWriteLock mu = new ReentrantReadWriteLock();
    private final Map<String, Topic> topics = new ConcurrentHashMap<>();
    private final Map<String, ConsumerGroup> consumerGroups = new ConcurrentHashMap<>();

    public void createTopic(String name, int numPartitions) {
        mu.writeLock().lock();
        try {
            if (topics.containsKey(name)) {
                throw new TopicExistsException();
            }
            Topic topic = new Topic(name, numPartitions, new KeyHashPartitioner());
            topics.put(name, topic);
        } finally {
            mu.writeLock().unlock();
        }
    }

    public ProduceResult produce(String topicName, String key, String value) {
        Topic topic;
        mu.readLock().lock();
        try {
            topic = topics.get(topicName);
        } finally {
            mu.readLock().unlock();
        }
        if (topic == null) {
            throw new TopicNotFoundException();
        }
        int partitionId = topic.selectPartition(key);
        long offset = topic.getPartitions().get(partitionId).append(key, value);
        return new ProduceResult(partitionId, offset);
    }

    private ConsumerGroup getOrCreateConsumerGroup(String groupId) {
        return consumerGroups.computeIfAbsent(groupId, ConsumerGroup::new);
    }

    /**
     * Returns up to maxMessages messages after the group's last committed
     * offset for (topic, partitionId), then auto-commits past the returned
     * batch. Auto-commit-on-read keeps this exercise's API small and
     * deterministic; the alternative is an explicit commitOffset call with
     * at-least-once redelivery on crash.
     */
    public List<Message> consume(String groupId, String topicName, int partitionId, int maxMessages) {
        Topic topic;
        mu.readLock().lock();
        try {
            topic = topics.get(topicName);
        } finally {
            mu.readLock().unlock();
        }
        if (topic == null) {
            throw new TopicNotFoundException();
        }
        if (partitionId < 0 || partitionId >= topic.getPartitions().size()) {
            throw new PartitionNotFoundException();
        }

        ConsumerGroup group = getOrCreateConsumerGroup(groupId);
        long fromOffset = group.committedOffset(topicName, partitionId);
        List<Message> messages = topic.getPartitions().get(partitionId).read(fromOffset, maxMessages);
        if (!messages.isEmpty()) {
            group.commit(topicName, partitionId, messages.get(messages.size() - 1).getOffset() + 1);
        }
        return messages;
    }

    public long committedOffset(String groupId, String topicName, int partitionId) {
        ConsumerGroup group = getOrCreateConsumerGroup(groupId);
        return group.committedOffset(topicName, partitionId);
    }
}
